extensions/bitwarden/src/session.tsx

```diff
@@ -0,0 +1,307 @@
+import { useVaultMessages } from "./hooks";
+import { pbkdf2 } from "crypto";
+import { Preferences } from "./types";
+
+const SessionContext = createContext<Session | null>(null);
+
+/**
+ * A Bitwarden login session.
+ */
+export interface Session {
+  readonly api: Bitwarden;
+
+  /**
+   * The session token.
+   */
+  readonly token: string | undefined;
+
+  readonly isLoading: boolean;
+  readonly isLocked: boolean;
+  readonly isAuthenticated: boolean;
+  readonly active: boolean;
+
+  readonly repromptHash: string | undefined;
+  readonly passwordEnteredDate: Date | undefined;
+
+  /**
+   * Locks the vault.
+   */
+  readonly lock: () => Promise<void>;
+
+  /**
+   * Logs out of the vault.
+   */
+  readonly logout: () => Promise<void>;
+
+  /**
+   * Checks if the password reprompt dialog can be skipped.
+   */
+  readonly canRepromptBeSkipped: () => boolean;
+
+  /**
+   * Checks if the provided password matches the master password.
+   */
+  readonly repromptConfirmation: (password: string) => Promise<boolean>;
+}
+
+/**
+ * Component which provides a session via the {@link useSession} hook.
+ *
+ * @param props.api The Bitwarden API.
+ * @param props.unlock If true, an unlock form will be displayed if the vault is locked or unauthenticated.
+ */
+export function SessionProvider(props: PropsWithChildren<{ api: Bitwarden; unlock?: boolean }>): JSX.Element {
+  const { api, children } = props;
+  const [session, setSession] = useState<Session>({
+    api,
+    token: undefined,
+    isLoading: true,
+    isLocked: false,
+    isAuthenticated: false,
+    active: false,
+
+    repromptHash: undefined,
+    passwordEnteredDate: undefined,
+
+    lock: async function lockVault() {
+      const toast = await showToast({ title: "Locking Vault...", style: Toast.Style.Animated });
+      await api.lock();
+      await deleteToken();
+      await toast.hide();
+    },
+
+    logout: async function logoutVault() {
+      const toast = await showToast({ title: "Logging Out...", style: Toast.Style.Animated });
+      await api.logout();
+      await deleteToken();
+      await toast.hide();
+    },
+
+    canRepromptBeSkipped: function (this: Session) {
+      const { repromptIgnoreDuration } = getPreferenceValues<Preferences>();
+      if (this.passwordEnteredDate == null) {
+        return false;
+      }
+
+      const skipDuration = parseInt(repromptIgnoreDuration, 10);
+      const skipUntil = this.passwordEnteredDate.getTime() + skipDuration;
+      return Date.now() <= skipUntil;
+    },
+
+    repromptConfirmation: async function (this: Session, password: string) {
+      const providedHashed = await hashMasterPasswordForReprompting(password);
+      if (providedHashed !== this.repromptHash) {
+        return false;
+      }
+
+      const now = new Date();
+      await LocalStorage.setItem(REPROMPT_PASSWORD_ENTERED_KEY, new Date().toString());
+      setSession((old) => ({ ...old, passwordEnteredDate: now }));
+      return true;
+    },
```

Comment:
I'm a bit confused with this approach. All these functions are being passed as the `useState`'s initial state, which means that they are not reactive to the component's scope, and also, they don't need to be changed programmatically, so no need for them to be here. I'm not sure if there's anything inside the functions that can change at runtime, but nonetheless, these should be moved outside the useState initializer, and passed to the provider below, alongside the rest of the values (e.g. token, isLoading, etc). Same for the `api` variable, which is already passed as prop.

Reply:
I wanted to make it a bit less verbose by just keeping everything within a single object and updating that, rather than having both a `Session` object for public API consumption and a separate `SessionState` object for holding the state data itself.

In hindsight, it was a pretty bad approach that made far less sense for the sake of brevity. I split them back up into a `Session` class and `SessionState` interface in my latest round of changes, and wired it up appropriately so the `Session` exposes only the necessary parts of the `SessionState`.

Reply:
Cool, nice work 👍 

I don't really like the class approach, this is React after all, and that could easily be handled by using hooks and some functions, but I'll leave that to be evaluated by @pomdtr.